exes/mqtt-broadcaster/inc/mqtt_broadcaster.hpp

```diff
@@ -0,0 +1,560 @@
+                                            async_mqtt::will(async_mqtt::allocate_buffer(will_topic),
+                                                             async_mqtt::buffer("DEATH"), async_mqtt::qos::at_least_once),
+                                            async_mqtt::allocate_buffer(config_.value()._mqtt_broker_username),
+                                            async_mqtt::allocate_buffer(config_.value()._mqtt_broker_password),
+                                            { async_mqtt::property::session_expiry_interval{ 0 } } };
+
+        logger_.trace("Sending MQTT connection packet...");
+
+        co_await mqtt_client_->send(connect_packet, asio::use_awaitable);
+
+        logger_.trace("MQTT connection packet sent. Waiting for CONNACK...");
+
+        auto connack_received = co_await mqtt_client_->recv(
+            async_mqtt::filter::match, { async_mqtt::control_packet_type::connack }, asio::use_awaitable);
+
+        auto connack_packet = connack_received.template get<async_mqtt::v5::connack_packet>();
+
+        if (connack_packet.code() != async_mqtt::connect_reason_code::success) {
+          logger_.error("Connection to MQTT broker failed with code: {}",
+                        async_mqtt::connect_reason_code_to_str(connack_packet.code()));
+
+          throw std::runtime_error("Connection to MQTT broker failed");
+        }
+
+        logger_.trace("Received CONNACK. Connection successful.");
+
+        co_return;
+      } catch (std::exception& e) {
+        logger_.error("Error while connecting to MQTT broker: {}", e.what());
+      }
+
+      logger_.trace("Waiting for 10 seconds and then trying to connect again...");
+      co_await asio::co_spawn(mqtt_client_->strand(), mqtt_client_->close(asio::use_awaitable), asio::use_awaitable);
+      co_await asio::steady_timer{ io_ctx_, std::chrono::seconds{ 10 } }.async_wait(asio::use_awaitable);
+    }
+  }
+
+  auto ncmd_listener() -> asio::awaitable<void> {
+    co_await subscribe_to_ncmd_topic();
+
+    while (true) {
+      try {
+        co_await process_ncmd_packet();
+      } catch (std::exception&) {
+        break;
+      }
+    }
+  }
+
+  auto process_ncmd_packet() -> asio::awaitable<void> {
+    try {
+      logger_.trace("Waiting for NCMD packet...");
+
+      auto publish_packet_received = co_await mqtt_client_->recv(
+          async_mqtt::filter::match, { async_mqtt::control_packet_type::publish }, asio::use_awaitable);
+
+      auto publish_packet = publish_packet_received.template get<async_mqtt::v5::publish_packet>();
+
+      logger_.trace("Received PUBLISH packet. Parsing payload...");
+
+      auto data = publish_packet.payload()[0];
+
+      Payload payload;
+
+      bool payload_valid = payload.ParseFromArray(data.data(), data.size());
+
+      if (!payload_valid) {
+        logger_.warn("Received invalid payload. Continuing to next iteration...");
```

Comment:
As taken from the relevant documentation:

```cpp
  // Fill the message with a protocol buffer parsed from the given input
  // stream. Returns false on a read error or if the input is in the wrong
  // format.  A successful return does not indicate the entire input is
  // consumed, ensure you call ConsumedEntireMessage() to check that if
  // applicable.
```

It will only return false on error, not the relevant error.